File: Cplusplus/Class_Demos/Nested.cpp
#include "stdafx.h"
#include <iostream>


/* 
 * NESTED CLASSES
 *
 * Inner class can access Outer's: 
 *		static vars, 
 *		private vars via pointer or reference
 *
 * 2 Inner classes can access each other's public fields
 * or access private fields if declared as 'friend' classes
 *
 * Outer class can access Inner's private members if it is 'friend'
 *
 * static vars must be defined outside class
 *		int Outer::Inner02::epoch = 1970;
 */

int glob;

class Outer {
	int d_value = 0;
	static int s_value;  // must be defined outside class to be visible

	class PrivateInner {
		int d_privateInner;
	// public members of private nested class
	// can be accessed from Outer class & other Inner classes
	public:
		PrivateInner() : d_privateInner(55) { }
		void privateInnerFunc() { std::cout << "HI!" << std::endl; }
	};

public:
	class Inner01 {
		class Inner02;	// forward declared Inner02 in Inner01
						// if contains pointers, references, parameters or 
						// return values to objects of the other nested classes
		void mutateOuter(Outer & o);
		Inner02 * ptr;
	};

	class Inner02 {
		friend class Outer;  // to access Inner's private fields

		int innerField;
	public:
		Inner02() : innerField(22) {
			std::cout << "Created Inner02 Object" << std::endl;
		}

		// No.1 reference to access Outer's private
		void setOuter1(Outer & outer, int i) {
			outer.d_value = i;
			std::cout << "Outer's d_value: " << outer.d_value << std::endl;

			s_value = i;
			std::cout << "Outer's s_value: " << s_value << std::endl;

			glob = i;
			std::cout << "Global's s_value: " << glob << std::endl;
		}

		// No.2 pointer to access Outer's private
		void setOuter2(Outer * p, int i) {
			p->d_value = i;
			std::cout << "Outer's d_value: " << p->d_value << std::endl;
		}

		Inner01 in1;	// Outer::Inner02 can access Outer::Inner01
		PrivateInner privateInnerObj; // Outer::Inner02 can access public fields Outer::PrivateInner
	};

	PrivateInner privateInnerObj;
	void Outer::caller() {
		privateInnerObj.privateInnerFunc();
	}

	static Inner02 getInner() { return Inner02(); }  // Outer class can use Inner as 'return' and 'args'

	int getInnerPrivate(Inner02 * p) {
		return p->innerField; 
	}
};

int Outer::s_value;  // static var definition

void Outer::Inner01::mutateOuter(Outer & o) {
	o.d_value++; 
}


int main() 
{
	Outer outer;
	Outer * outerPtr = new Outer;
	Outer::Inner02 inner;
	Outer::Inner02 * innerPtr = new Outer::Inner02;

	inner.setOuter1(outer, 10);		// pass reference to Outer
	inner.setOuter2(outerPtr, 100);	// pass pointer to Outer

	outer.caller();
	outerPtr->caller();

	Outer::Inner02 inner2 = outer.getInner();
	auto a = outer.getInnerPrivate(innerPtr);
	std::cout << "Inner's member: " << a << std::endl;


	std::cin.get();
}

File: Csharp/Nested.cs
using System;
using System.Runtime.InteropServices;


namespace CSharpApp
{
    // use a static class for 'globals'
    public static class MyGlobals
    {
        public static int glob; 
    }
    class Outer
    {
        private int d_value = 0;
        private static int s_value; 
        private class PrivateInner
        {
            private int d_privateInner;
		    public PrivateInner() 
            {
                d_privateInner = 55;
            }
            public void privateInnerFunc() { Console.WriteLine("PrivateInner::privateInnerFunc()"); }
        }
        public class Inner01
        {
            private void mutateOuter(ref Outer oc)
            {
                oc.d_value++;
            }
        }
        public class Inner02
        {
            internal int innerField;    // use internal instead of private
                                        // acsess level compares with 'friend' class in C++
            public Inner02()
            {
                innerField = 22;
                Console.WriteLine("Created Inner02 Object");
            }
            public void setOuter(ref Outer oc, int i)
            {
                Outer.s_value = i;
                Console.WriteLine($"Outer's s_value: { Outer.s_value }");

                oc.d_value = i;
                Console.WriteLine($"Outer's d_value: { oc.d_value }");

                MyGlobals.glob = i;
                Console.WriteLine($"Global's s_value: { MyGlobals.glob }");
            }
            Inner01 in1 = new Inner01();    // Outer::Inner02 can access Outer::Inner01
            PrivateInner privateInnerObj = new PrivateInner();  // Outer::Inner02 can access public fields Outer::PrivateInner
        }

        PrivateInner privateInnerObj = new PrivateInner();  // Outer creates object of PrivateInner
        public void caller()
        {
            privateInnerObj.privateInnerFunc();
        }
        public static Inner01 getInner() { return new Inner01(); }  // Outer class can use Inner as 'return' and 'args'
        public int getInnerPrivate(ref Inner02 ic)
        {
            return ic.innerField;
        }
    }

    class Demo
    {
        static void Main(string[] args)
        {
            Outer outer = new Outer();
            Outer.Inner02 inner2 = new Outer.Inner02();

            inner2.setOuter(ref outer, 10);		// pass reference to Outer

            outer.caller();

            Outer.Inner01 inner1 = Outer.getInner();

            var a = outer.getInnerPrivate(ref inner2);
            Console.WriteLine($"Inner's member: { a }");


            Console.Read();
        }
    }
}
